packages/cli/src/routeGeneration/templates/express.hbs

```diff
@@ -104,51 +104,73 @@ export function RegisterRoutes(app: express.Router) {
+        return reverse(Promise.all([...iterable].map(reverse)));
+    };
+
+    // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
+
+    function authenticateMiddleware(security: TsoaRoute.Security[] = []) {
+        return async function runAuthenticationMiddleware(request: any, _response: any, next: any) {
 
             // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
+            // keep track of failed auth attempts so we can hand back the most
+            // recent one.  This behavior was previously existing so preserving it
+            // here
+            const failedAttempts: any[] = [];
+            const pushAndRethrow = (error: any) => {
+                failedAttempts.push(error);
+                throw error;
+            };
+
+            const secMethodOrPromises: Promise<any>[] = [];
             for (const secMethod of security) {
                 if (Object.keys(secMethod).length > 1) {
-                    let promises: Promise<any>[] = [];
+                    const secMethodAndPromises: Promise<any>[] = [];
 
                     for (const name in secMethod) {
-                        promises.push(expressAuthentication(request, name, secMethod[name]));
+                        secMethodAndPromises.push(
+                            expressAuthentication(request, name, secMethod[name])
+                                .catch(pushAndRethrow)
+                        );
                     }
 
-                    Promise.all(promises)
-                        .then((users) => { succeed(users[0]); })
-                        .catch(fail);
+                    // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
+
+                    secMethodOrPromises.push(Promise.all(secMethodAndPromises)
+                        .then(users => { return users[0]; }));
                 } else {
                     for (const name in secMethod) {
-                        expressAuthentication(request, name, secMethod[name])
-                            .then(succeed)
-                            .catch(fail);
+                        secMethodOrPromises.push(
+                            expressAuthentication(request, name, secMethod[name])
+                                .catch(pushAndRethrow)
+                        );
                     }
                 }
             }
+
+            // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
+
+            try {
+                request['user'] = await promiseAny(secMethodOrPromises);
+                next();
+            }
+            catch(err) {
```

Comment:
Looking at https://github.com/lukeautry/tsoa/issues/659 here.
The Promise.any would throw an AggregateError here, which would be great for determining what the respective issues were.
However the current Polyfill doesn't do that. Wdyt about a spec compliant shim?

Reply:
This would be a breaking change, but one that makes sense and allows us to eventually remove witout headaches.

Reply:
I'm all for bringing in a spec compliant shim. It would utilize shared code better by using a single import and it would be easy to pull it out later once Promise.any becomes commonplace.

I'm not convinced that we'd want to return the aggregate error here however.  If we returned the AggregateError we'd still need some code outside of the framework to determine which one is the correct one to show the user.  It feels like integrating that into the authorization code is the way to go since it already exists and its a singular spot to contain crucial details that may not be passed along in the AggregateError.  Here's one example.

say we're using two kinds of auth:
1) bearer token authentication
2) api key

A user with a bearer token which is a JWT attempts to authenticate but the token is expired

the code checks bearer token auth and throws an Error saying "Token Expired"
the code then checks the api key auth and see's that there is no X-Api-Key Header at all so throws an error saying "Missing X-Api-Key Header". 

The current templates would return the error saying "missing X-Api-Key Header" since it was last to fail.  Inside the authorization code we've just checked for the X-Api-Key header, seen that its missing and we have a pretty good idea that this type of authentication is not the intended one.  If we sent both errors inside of an AggregateError we'd still need to do a follow up check on the request object to see that X-Api-Key header is missing.  This bit of code would probably need to do a lot of little checks which would end up looking like the authorization code path already does.

Reply:
I did push a fix up to show what the changes look like to use the polyfill instead.  I ended up using the imported method directly instead of executing any.shim() because typescript still didn't know about `Promise.any` after the shim ran so it would still involve extra changes to suppress this.

Reply:
The imported method is perfectly fine.
For your example I'd argue it should be up to the developer to decide which error to show.

> still need some code outside of the framework to determine which one is the correct one to show the user

We already pass down the error, so this would not change. However, we could let the developer decide which Error to show if multiple errors occurred (in your example, we might end up passing down the Error for 2) when the token is expired.

> The current templates would return the error saying "missing X-Api-Key Header" since it was last to fail. 

Which is obviously what I'd want to prevent. Expired Token is the correct issue to face, wouldn't you agree?

Instead of one error you can have an AggregateError you can use to derive what the issues were.

Reply:
I agree we need to have the developer decide which error to show.  I'm seeing a few possibilities for how this could work.

1) In the event that no Authorization methods are accepted, we throw the aggregate error that contains each error.  In order to determine which error to show, the developer would need to add an additional middleware which runs after the generated routes to intercept the error and choose which one to show. In order to figure out the correct error, the developer would need to write a set of checks for each auth type.  For example, If X-api-key header was not passed, we know the authentication isn't any that relied on that header, etc.  These are probably simple checks depending on what the authentication methods are and how many there are.  Adding a new authentication method means updating the authentication.ts file and updating the middleware which is responsible to determine how to pick the correct error out of the AggregateError.  This would be a breaking change for users with multiple types of authentication.

2) We allow the authentication methods for each type of auth pass a special value in the thrown error which helps determine whether this is the correct error to show.  For example, this could be a priority value as mentioned in  #659  or it could be something like a flag which explicitly removes the error from consideration. For example `error.skip = true`.  The routes file then uses these additional flags to figure out the correct error and then falls back to existing behavior if the values are ambiguous or don't exist.  In this case, adding a new authentication method means just updating the authentication.ts file and not a new middleware.  It is not a breaking change since it falls back to existing behavior.  There is an existing pattern of adding special values to the error, see here for customizing status code via the authentication function: https://github.com/lukeautry/tsoa/blob/master/packages/cli/src/routeGeneration/templates/express.hbs#L126

3) we could add an additional optional developer defined function similar to the auth methods in the authentication.ts file.  It could be a specifically named method that is exported from authentication.ts which is specifically for determining which error to choose in this case.  If that method does not exist TSOA could fall back to the existing behavior of choosing the most recently failed.  This would avoid the breaking change associated with option 1.

I personally prefer option 2 because it doesn't require a breaking change and because i think that the code which is required to check whether the auth method is appropriate looks a lot like the code which is already in the authentication implementation. My assumption is that its just a line or two of code for each method to return whether its one you want to show.

Ultimately, I'll leave it up to you @WoH to decide which one you think is best for the framework.  Let me know your thoughts.

Reply:
One more option is keep what I have already which preserves the existing behavior and consider the other changes as part of a separate PR.  That would let us get this landed sooner.

Reply:
I hope you're not making this assumption based on my (lack) of quick feedback which should be attributed to my schedule, I am sorry about that.

Reply:
If you feel like making a non-breaking change and expanding on it with a (smaller) breaking change, I'm happy to do that, up to you. My thought was to make it easier on you by making only one PR and merging it with 2 other breaking changes that I have on mind for the next release (TS String Literals, node > 12, #997, possibly #1008)

Reply:
The problem with 2) is I don't see how the priority would solve the issue. I.e. if there is no Authorization header that's low prio. If the Authorization header has an expired token, that's now very high priority for me, since I'm expecing the client intended to use that token mechanism to authorize. But it's the same token strategy in both cases.

Reply:
If you want to go with option 1 and make the breaking change, it would be good to get it out with  the next release.  Is that coming up soon?

I am thinking that it would be better to make a separate PR for the altering of error return behavior. It will be easier to understand from a historical perspective to keep the 2 issues separate.

Are you okay with this one as is considering it preserves existing behavior?  I've gone ahead and rebased in the fixup to use the external promise definition.

Reply:
Hmm, I just did some local testing using the updated version of this branch and ran into a couple issues.
1) my local project is complaining that the types for promise.any weren't installed. 
2) There's something weird about how the promise.any stuff is exported.  My project is complaining about the import * as promiseAny syntax. I was able to get it working 2 ways, I'm wondering which you recommend or if something else should be done:
These seem to work:
const promiseAny = require('promise.any');  // not ideal because it uses require, do some typescript configs restrict that?

Reply:
@WoH do you think the @types/promise.any would need to be included as a full dependency?  I don't think subpackage devDependencies are installed and if the developer using noImplictAny, they'll get a type error.

Reply:
TS-ignore them or cast to explicit any

Reply:
okay, ts-ignore works well for the lack of types.  I ended up having to use the require for this shim because i couldn't get the tests working without it.  Not sure why.

Reply:
Can you elaborate? 
```
import * as promiseAny from 'promise.any'
```

should be fine.

Reply:
I retried again and it seems to be working fine.  I think my environment was slightly messed up.  I'll push that change back up.

Reply:
actually it doesn't work. This version builds and passes the tests but when I test it with my project I get this error:

`err:  TypeError: promiseAny is not a function`

I looked at the promise.any code and its doing something fancy to make the import work as a function an object. I'm wondering if this is causing it to break using this syntax.  I'm unclear why its not breaking for the tests.  Maybe I've got my project configured to handle the imports differently?

here is the source:
https://github.com/es-shims/Promise.any/blob/master/index.js